### PR TITLE
Release v1.12.1

### PR DESCRIPTION
**Pull request by EmilyBourne**
Create a patch release including the following commits from the devel branch:

- [x] d64be7ff7 (improved print_constant (#1120), 2024-05-14)
- [x] 28f555d3c (Store annotated local variables to avoid memory leaks (#1904), 2024-05-29)
- [x] d7c5f3500 (Improve Error Message for Missing Target Language Compiler in Pyccel (#1928), 2024-07-10)
- [x] e4dfd4f8d (Fix code exportation (#1934), 2024-07-17)
- [x] 68532aaf2 (Correct results type for untransalatable function (#1950), 2024-07-26)
- [x] f77d57f8f (Fix usage of used_names in pyccel_to_sympy (#1958), 2024-08-16)
- [x] 41093a8b3 (Ensure an error is raised when passing the wrong type as an optional argument (#1971), 2024-09-03)
- [x] 6f2c47018 (Add NumPy v2 compatible sign function (#1984), 2024-09-18)
- [x] dbcbf5823 (fix gcd and lcm implementations (#1986), 2024-09-22)
- [x] 32ee7ee93 (Add support for NumPy v2 (#1988), 2024-09-24)
- [x] d2ca21f48f0a94c7ff2a7314ea4235a59948cfac (Set up GitHub release (#1997), 2024-09-27)

Additionally the following PR was required to handle NumPy update problems with the old typing system:
- [x] e17c366a26b6d7028e365c04f4a46e7b58ee6ad8 (Fix type bugs with NumPy v2 (#1993), 2024-09-25)

### CODE REVIEW

File: pyccel/codegen/printing/ccode.py [hunk 1]
```diff
@@ -2290,14 +2268,19 @@ def _print_Constant(self, expr):
...
+            return 'NAN'
         elif expr == math_constants['pi']:
             self.add_import(c_imports['math'])
             return 'M_PI'
         elif expr == math_constants['e']:
             self.add_import(c_imports['math'])
             return 'M_E'
         else:
-            raise NotImplementedError("Constant not implemented")
+            cast_func = DtypePrecisionToCastFunction[expr.dtype]
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:2281]: /bot accept
This is a fallback solution which was accepted in the associated PR

File: pyccel/ast/internals.py [hunk 2]
```diff
@@ -409,7 +413,10 @@ def max_precision(objs : list, allow_native : bool = True):
         ndarray_list = [o for o in objs if getattr(o, 'is_ndarray', False)]
         if ndarray_list:
             return get_final_precision(max(ndarray_list, key=attrgetter('precision')))
-        return max(get_final_precision(o) for o in objs)
+        if numpy_v1:
+            return max(get_final_precision(o) for o in objs)
+        else:
+            return get_final_precision(max(objs, key = lambda o : o.precision))
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/internals.py:419]: /bot accept
Tested by macosx

File: pyccel/ast/numpyext.py [hunk 3]
```diff
@@ -1922,8 +1924,11 @@
         x : TypedAstNode
             The argument passed to the function.
         """
-        self._dtype     = NativeFloat()
-        self._precision = default_precision[self._dtype]
+        if numpy_v1:
+            super()._set_dtype_precision(x)
+        else:
+            self._dtype = x.dtype
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/numpyext.py:1930]: /bot accept
Tested by macosx

**Review by yguclu — APPROVED** [review]

**Pull request merged**